File: sketch/src/main/java/me/xiaopan/sketch/decode/ThumbnailModeDecodeHelper.java
/*
 * Copyright (C) 2017 Peng fei Pan <sky@xiaopan.me>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xiaopan.sketch.decode;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.Build;

import me.xiaopan.sketch.ErrorTracker;
import me.xiaopan.sketch.SLog;
import me.xiaopan.sketch.SLogType;
import me.xiaopan.sketch.cache.BitmapPool;
import me.xiaopan.sketch.cache.BitmapPoolUtils;
import me.xiaopan.sketch.drawable.ImageAttrs;
import me.xiaopan.sketch.request.LoadOptions;
import me.xiaopan.sketch.request.LoadRequest;
import me.xiaopan.sketch.request.Resize;
import me.xiaopan.sketch.util.SketchUtils;

public class ThumbnailModeDecodeHelper extends DecodeHelper {
    private static final String LOG_NAME = "ThumbnailModeDecodeHelper";

    /**
     * 要想使用缩略图功能需要配置开启缩略图功能、配置resize并且图片格式和系统版本支持BitmapRegionDecoder才行
     */
    @Override
    public boolean match(LoadRequest request, DataSource dataSource, ImageType imageType, BitmapFactory.Options boundOptions) {
        LoadOptions loadOptions = request.getOptions();
        if (!loadOptions.isThumbnailMode()
                || !SketchUtils.sdkSupportBitmapRegionDecoder()
                || !SketchUtils.formatSupportBitmapRegionDecoder(imageType)) {
            return false;
        }

        Resize resize = loadOptions.getResize();
        if (resize == null) {
            SLog.e(LOG_NAME, "thumbnailMode need resize ");
            return false;
        }

        // 只有原始图片的宽高比和resize的宽高比相差3倍的时候才能使用缩略图方式读取图片
        ImageSizeCalculator sizeCalculator = request.getConfiguration().getImageSizeCalculator();
        return sizeCalculator.canUseThumbnailMode(boundOptions.outWidth, boundOptions.outHeight,
                resize.getWidth(), resize.getHeight());
    }

    @Override
    public DecodeResult decode(LoadRequest request, DataSource dataSource, ImageType imageType, BitmapFactory.Options boundOptions,
                               BitmapFactory.Options decodeOptions, int exifOrientation) throws DecodeException {

        ImageOrientationCorrector orientationCorrector = request.getConfiguration().getImageOrientationCorrector();
        orientationCorrector.rotateSize(boundOptions, exifOrientation);

        // 缩略图模式强制质量优先
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD_MR1
                && !decodeOptions.inPreferQualityOverSpeed) {
            decodeOptions.inPreferQualityOverSpeed = true;
        }

        // 计算resize区域在原图中的对应区域
        LoadOptions loadOptions = request.getOptions();
        Resize resize = loadOptions.getResize();
        ResizeCalculator resizeCalculator = request.getConfiguration().getResizeCalculator();
        ResizeCalculator.Mapping mapping = resizeCalculator.calculator(boundOptions.outWidth, boundOptions.outHeight,
                resize.getWidth(), resize.getHeight(), resize.getScaleType(), false);

        boolean supportLargeImage = SketchUtils.supportLargeImage(request, imageType);

        // 根据resize的大小和原图中对应区域的大小计算缩小倍数，这样会得到一个较为清晰的缩略图
        ImageSizeCalculator sizeCalculator = request.getConfiguration().getImageSizeCalculator();
        decodeOptions.inSampleSize = sizeCalculator.calculateInSampleSize(mapping.srcRect.width(), mapping.srcRect.height(),
                resize.getWidth(), resize.getHeight(), supportLargeImage);

        orientationCorrector.reverseRotate(mapping.srcRect, boundOptions.outWidth, boundOptions.outHeight, exifOrientation);

        if (BitmapPoolUtils.sdkSupportInBitmapForRegionDecoder() && !loadOptions.isBitmapPoolDisabled()) {
            BitmapPool bitmapPool = request.getConfiguration().getBitmapPool();
            BitmapPoolUtils.setInBitmapFromPoolForRegionDecoder(decodeOptions, mapping.srcRect, bitmapPool);
        }

        Bitmap bitmap = null;
        try {
            bitmap = ImageDecodeUtils.decodeRegionBitmap(dataSource, mapping.srcRect, decodeOptions);
        } catch (Throwable throwable) {
            throwable.printStackTrace();

            ErrorTracker errorTracker = request.getConfiguration().getErrorTracker();
            BitmapPool bitmapPool = request.getConfiguration().getBitmapPool();
            if (ImageDecodeUtils.isInBitmapDecodeError(throwable, decodeOptions, true)) {
                ImageDecodeUtils.recycleInBitmapOnDecodeError(errorTracker, bitmapPool, request.getUri(),
                        boundOptions.outWidth, boundOptions.outHeight, boundOptions.outMimeType, throwable, decodeOptions, true);

                try {
                    bitmap = ImageDecodeUtils.decodeRegionBitmap(dataSource, mapping.srcRect, decodeOptions);
                } catch (Throwable throwable1) {
                    throwable1.printStackTrace();

                    errorTracker.onDecodeNormalImageError(throwable1, request, boundOptions.outWidth,
                            boundOptions.outHeight, boundOptions.outMimeType);
                }
            } else if (ImageDecodeUtils.isSrcRectDecodeError(throwable, boundOptions.outWidth, boundOptions.outHeight, mapping.srcRect)) {
                errorTracker.onDecodeRegionError(request.getUri(), boundOptions.outWidth, boundOptions.outHeight,
                        boundOptions.outMimeType, throwable, mapping.srcRect, decodeOptions.inSampleSize);
            } else {
                errorTracker.onDecodeNormalImageError(throwable, request, boundOptions.outWidth,
                        boundOptions.outHeight, boundOptions.outMimeType);
            }
        }

        // 过滤掉无效的图片
        if (bitmap == null || bitmap.isRecycled()) {
            ImageDecodeUtils.decodeError(request, dataSource, LOG_NAME);
            return null;
        }

        // 过滤宽高小于等于1的图片
        if (bitmap.getWidth() <= 1 || bitmap.getHeight() <= 1) {
            if (SLogType.REQUEST.isEnabled()) {
                SLog.fw(SLogType.REQUEST, LOG_NAME,
                        "image width or height less than or equal to 1px. imageSize: %dx%d. bitmapSize: %dx%d. %s",
                        boundOptions.outWidth, boundOptions.outHeight, bitmap.getWidth(), bitmap.getHeight(), request.getKey());
            }
            bitmap.recycle();
            ImageDecodeUtils.decodeError(request, dataSource, LOG_NAME);
            return null;
        }

        ImageAttrs imageAttrs = new ImageAttrs(boundOptions.outMimeType, boundOptions.outWidth, boundOptions.outHeight, exifOrientation);
        BitmapDecodeResult result = new BitmapDecodeResult(imageAttrs, bitmap).setProcessed(true);

        correctOrientation(orientationCorrector, result, exifOrientation, request);

        ImageDecodeUtils.decodeSuccess(bitmap, boundOptions.outWidth, boundOptions.outHeight, decodeOptions.inSampleSize, request, LOG_NAME);
        return result;
    }
}

File: sketch3/src/main/java/com/github/panpf/sketch/common/decode/internal/ThumbnailModeDecodeHelper.kt
/*
 * Copyright (C) 2019 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.common.decode.internal

import android.graphics.Bitmap
import android.graphics.BitmapFactory.Options
import android.graphics.BitmapRegionDecoder
import android.graphics.Point
import android.os.Build
import android.os.Build.VERSION_CODES
import com.github.panpf.sketch.SLog
import com.github.panpf.sketch.Sketch
import com.github.panpf.sketch.common.DecodeException
import com.github.panpf.sketch.common.ImageType
import com.github.panpf.sketch.common.LoadableRequest
import com.github.panpf.sketch.common.cache.BitmapPoolHelper
import com.github.panpf.sketch.common.datasource.DataSource
import com.github.panpf.sketch.common.decode.DecodeResult
import com.github.panpf.sketch.common.decode.internal.ResizeCalculator.Mapping
import com.github.panpf.sketch.load.ImageInfo
import com.github.panpf.sketch.load.Resize
import com.github.panpf.sketch.util.supportBitmapRegionDecoder

/**
 * 缩略图模式解码协助器，当开启缩略图模式并且满足使用缩略图模式的条件时会使用此协助器来解码
 *
 * 解码时会根据 resize 的尺寸并使用 [BitmapRegionDecoder] 读取原图中的部分区域来得到更清晰的缩略图
 */
class ThumbnailModeDecodeHelper {

    companion object {
        private const val MODULE = "ThumbnailModeDecodeHelper"
    }

    private val resizeCalculator = ResizeCalculator()
    private val sizeCalculator = ImageSizeCalculator()

    fun canUseThumbnailMode(
        resize: Resize?, imageInfo: ImageInfo, imageType: ImageType?
    ): Boolean {
        if (resize != null && resize.thumbnailMode && imageType?.supportBitmapRegionDecoder() == true) {
            if (resize.width <= imageInfo.width || resize.height <= imageInfo.height) {
                val resizeScale = resize.width.toFloat() / resize.height
                val imageScale = imageInfo.width.toFloat() / imageInfo.height
                return resizeScale.coerceAtLeast(imageScale) > resizeScale.coerceAtMost(imageScale) * 1.5f
            }
        }
        return false
    }

    fun decode(
        sketch: Sketch,
        request: LoadableRequest,
        dataSource: DataSource,
        imageInfo: ImageInfo,
        decodeOptions: Options,
    ): DecodeResult {
        val resize = request.resize!!
        val maxSize = request.maxSize
        val exifOrientation = imageInfo.exifOrientation
        val bitmapPoolHelper = sketch.bitmapPoolHelper

        if (Build.VERSION.SDK_INT <= VERSION_CODES.M && !decodeOptions.inPreferQualityOverSpeed) {
            decodeOptions.inPreferQualityOverSpeed = true
        }

        val imageSize = Point(imageInfo.width, imageInfo.height)

        val imageOrientationCorrector =
            ImageOrientationCorrector.fromExifOrientation(exifOrientation)
        imageOrientationCorrector?.rotateSize(imageSize)

        val resizeMapping = resizeCalculator.calculator(
            imageSize.x,
            imageSize.y,
            resize.width,
            resize.height,
            resize.scaleType,
            false
        )
        val resizeMappingSrcWidth = resizeMapping.srcRect.width()
        val resizeMappingSrcHeight = resizeMapping.srcRect.height()

        val resizeInSampleSize =
            sizeCalculator.calculateInSampleSize(
                resizeMappingSrcWidth, resizeMappingSrcHeight, resize.width, resize.height
            )
        val maxSizeInSampleSize = maxSize?.let {
            sizeCalculator.calculateInSampleSize(
                resizeMappingSrcWidth, resizeMappingSrcHeight, it.width, it.height
            )
        } ?: 1
        decodeOptions.inSampleSize = resizeInSampleSize.coerceAtLeast(maxSizeInSampleSize)

        imageOrientationCorrector?.reverseRotateRect(
            resizeMapping.srcRect,
            imageSize.x,
            imageSize.y,
        )

        if (request.disabledBitmapPool != true) {
            sketch.bitmapPoolHelper.setInBitmapForRegionDecoder(
                decodeOptions,
                resizeMapping.srcRect
            )
        }

        val bitmap = decodeRegionBitmap(
            request, dataSource, resizeMapping, decodeOptions,
            bitmapPoolHelper, imageSize, imageInfo
        )

        val correctedOrientationBitmap =
            imageOrientationCorrector?.rotateBitmap(bitmap, bitmapPoolHelper.bitmapPool)
        return if (correctedOrientationBitmap != null && correctedOrientationBitmap != bitmap) {
            bitmapPoolHelper.freeBitmapToPool(bitmap)
            DecodeResult(correctedOrientationBitmap, imageInfo)
        } else {
            DecodeResult(bitmap, imageInfo)
        }
    }

    private fun decodeRegionBitmap(
        request: LoadableRequest,
        dataSource: DataSource,
        mapping: Mapping,
        decodeOptions: Options,
        bitmapPoolHelper: BitmapPoolHelper,
        imageSize: Point,
        imageInfo: ImageInfo,
    ): Bitmap {
        val bitmap = try {
            dataSource.decodeRegionBitmap(mapping.srcRect, decodeOptions)
        } catch (throwable: Throwable) {
            throwable.printStackTrace()
            when {
                isInBitmapError(throwable, decodeOptions, true) -> {
                    val message = "Bitmap region decode error. Because inBitmap. uri=%s"
                        .format(request.uri)
                    SLog.emt(MODULE, throwable, message)

                    val inBitmap = decodeOptions.inBitmap
                    decodeOptions.inBitmap = null
                    bitmapPoolHelper.freeBitmapToPool(inBitmap)
                    try {
                        dataSource.decodeRegionBitmap(mapping.srcRect, decodeOptions)
                    } catch (throwable2: Throwable) {
                        throwable.printStackTrace()
                        val message2 = "Bitmap region decode error. uri=%s".format(request.uri)
                        SLog.emt(MODULE, throwable2, message2)
                        throw DecodeException(message2, throwable2)
                    }
                }
                isSrcRectError(throwable, imageSize.x, imageSize.y, mapping.srcRect) -> {
                    val message =
                        "Bitmap region decode error. Because srcRect. imageInfo=%s, resize=%s, srcRect=%s, uri=%s"
                            .format(imageInfo, request.resize, mapping.srcRect, request.uri)
                    SLog.emt(MODULE, throwable, message)
                    throw DecodeException(message, throwable)
                }
                else -> {
                    val message = "Bitmap region decode error. uri=%s".format(request.uri)
                    SLog.emt(MODULE, throwable, message)
                    throw DecodeException(message, throwable)
                }
            }
        }
        if (bitmap == null) {
            val message = "Bitmap region decode return null. uri=%s".format(request.uri)
            SLog.em(MODULE, message)
            throw DecodeException(message)
        }

        if (bitmap.width <= 1 || bitmap.height <= 1) {
            bitmap.recycle()
            val message = "Invalid image size. size=%dx%d, uri=%s".format(
                imageInfo.width,
                imageInfo.height,
                request.uri
            )
            SLog.em(BitmapFactoryDecoder.MODULE, message)
            throw DecodeException(message)
        }

        return bitmap
    }
}